### PR TITLE
ref(useProject): Create helper useProject() to replace useProjects()

### PR DESCRIPTION
**Pull request by ryan953**
Relates to https://github.com/getsentry/sentry/issues/65949

### CODE REVIEW

File: static/app/utils/project/useAllProjectVisibility.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  };
+}
+
+describe('useAllProjectVisibility', () => {
+  const mockOrg = OrganizationFixture();
+  jest.mocked(useOrganization).mockReturnValue(mockOrg);
+
+  const MOCK_API_ENDPOINT = `/organizations/${mockOrg.slug}/projects/`;
+
+  const project10 = ProjectFixture({id: '10', slug: 'ten'});
```
Comment (scttcper) [inline static/app/utils/project/useAllProjectVisibility.spec.tsx:29]: might want to make a ProjectVisibilityFixture that just calls projectToProjectVisibility for you
```suggestion
  const project10 = ProjectFixture({id: '10', slug: 'ten'});
```

File: static/app/utils/project/useAllProjectVisibility.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      pages.flatMap((items): ProjectVisibility[] =>
+        items.map(projectToProjectVisibility)
+      ),
+    [pages]
+  );
+
+  const getBySlug = useCallback(
+    (slug: string | undefined): ProjectVisibility | undefined =>
+      slug
+        ? Object.fromEntries(projects.map(project => [project.slug, project]))[slug]
```
Comment (scttcper) [inline static/app/utils/project/useAllProjectVisibility.tsx:52]: should this be a .find?
```suggestion
      slug
        ? projects.find(project => project.slug === slug)
```

File: static/app/utils/project/useProject.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import type {ApiResult} from 'sentry/api';
+import type {Project} from 'sentry/types';
+import useAggregatedQueryKeys from 'sentry/utils/api/useAggregatedQueryKeys';
+import useAllProjectVisibility from 'sentry/utils/project/useAllProjectVisibility';
+import type {ApiQueryKey} from 'sentry/utils/queryClient';
+import useOrganization from 'sentry/utils/useOrganization';
+
+type Props =
+  | {slug: string | undefined; id?: undefined}
+  | {id: string | undefined; slug?: undefined};
```
Comment (scttcper) [inline static/app/utils/project/useProject.tsx:12]: i know you're avoiding checking if the object contains the keys, could use never too
```suggestion
  | {slug: string | undefined; id?: never}
  | {id: string | undefined; slug?: never};
```

File: static/app/utils/project/useProject.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    response: ApiResult,
+    _aggregates: readonly AggQueryKey[]
+  ) => ({
+    ...prevState,
+    ...Object.fromEntries(response[0].map(project => [project[fieldName], project])),
+  });
+}
+
+export default function useProject({slug, id}: Props) {
+  const organiation = useOrganization();
```
Comment (davidenwang) [inline static/app/utils/project/useProject.tsx:30]: nit: organization*

**Pull request closed without merging**